File: backend/apps/media_library/schemas.py
from datetime import datetime
from uuid import UUID

from ninja import Schema
from pydantic import Field


class MediaFolderCreateSchema(Schema):
    name: str = Field(min_length=1, max_length=200)
    parent_id: UUID | None = None
    description: str = ""


class MediaFolderSchema(Schema):
    id: UUID
    name: str
    slug: str
    parent_id: UUID | None
    description: str
    created_at: datetime
    updated_at: datetime


class MediaAssetUpdateSchema(Schema):
    folder_id: UUID | None = None
    title: str = Field(min_length=1, max_length=250)
    media_type: str
    alt_text: str = Field(default="", max_length=250)
    caption: str = ""
    description: str = ""
    tags: list[str] = Field(default_factory=list)
    is_public: bool = True


class MediaUsageCreateSchema(Schema):
    application: str = Field(min_length=1, max_length=100)
    model_name: str = Field(min_length=1, max_length=100)
    object_id: str = Field(min_length=1, max_length=100)
    field_name: str = Field(default="", max_length=100)
    usage_context: str = Field(
        default="",
        max_length=200,
    )


class MediaUsageSchema(Schema):
    id: UUID
    application: str
    model_name: str
    object_id: str
    field_name: str
    usage_context: str
    created_at: datetime


class MediaAssetSchema(Schema):
    id: UUID
    folder_id: UUID | None
    folder_name: str | None
    title: str
    file_url: str
    original_name: str
    media_type: str
    mime_type: str
    extension: str
    size: int
    width: int | None
    height: int | None
    duration_seconds: int | None
    alt_text: str
    caption: str
    description: str
    tags: list[str]
    checksum: str
    is_optimized: bool
    optimized_file_url: str | None
    webp_file_url: str | None
    is_public: bool
    usages: list[MediaUsageSchema]
    created_at: datetime
    updated_at: datetime

File: apps/dashboard/src/features/media-library/schemas.ts
import {
  z,
} from "zod";

import {
  mediaTypes,
} from "./types";

export const mediaFolderSchema =
  z.object({
    id: z.string().uuid(),
    name: z.string(),
    slug: z.string(),
    parent_id:
      z.string().uuid().nullable(),
    description: z.string(),
    created_at: z.string(),
    updated_at: z.string(),
  });

export const mediaUsageSchema =
  z.object({
    id: z.string().uuid(),
    application: z.string(),
    model_name: z.string(),
    object_id: z.string(),
    field_name: z.string(),
    usage_context: z.string(),
    created_at: z.string(),
  });

export const mediaAssetSchema =
  z.object({
    id: z.string().uuid(),
    folder_id:
      z.string().uuid().nullable(),
    folder_name:
      z.string().nullable(),
    title: z.string(),
    file_url: z.string(),
    original_name: z.string(),
    media_type: z.enum(mediaTypes),
    mime_type: z.string(),
    extension: z.string(),
    size:
      z.number().int().nonnegative(),
    width:
      z.number().int().nonnegative().nullable(),
    height:
      z.number().int().nonnegative().nullable(),
    duration_seconds:
      z.number().int().nonnegative().nullable(),
    alt_text: z.string(),
    caption: z.string(),
    description: z.string(),
    tags: z.array(z.string()),
    checksum: z.string(),
    is_optimized: z.boolean(),
    optimized_file_url:
      z.string().nullable(),
    webp_file_url:
      z.string().nullable(),
    is_public: z.boolean(),
    usages:
      z.array(mediaUsageSchema),
    created_at: z.string(),
    updated_at: z.string(),
  });

export const mediaFolderListSchema =
  z.array(mediaFolderSchema);

export const paginatedMediaAssetsSchema =
  z.object({
    items: z.array(mediaAssetSchema),
    pagination: z.object({
      page:
        z.number().int().positive(),
      page_size:
        z.number().int().positive(),
      total_items:
        z.number().int().nonnegative(),
      total_pages:
        z.number().int().nonnegative(),
    }),
  });

export const messageSchema =
  z.object({
    status: z.string(),
    message: z.string(),
  });
